### ISSUE TITLE
Cannot register Tart VMs as developer device to provisioning profile

### ISSUE THREAD

**Issue by s6eskand** [cirruslabs/tart#943]
Hello,

I am trying to build a macos app from src on a tart VM running macOS sequoia, and in order to be able to, the provisioning profile needs to have the device (tart VM) included. The device ID here is the mac UDID except I am getting the following error when attempting the register the device

```
{
  "errors" : [ {
    "id" : "<id>",
    "status" : "409",
    "code" : "ENTITY_ERROR.ATTRIBUTE.INVALID",
    "title" : "An attribute in the provided entity has invalid value",
    "detail" : "An invalid value '<tart_vm_udid>' was provided for the parameter 'udid'."
  } ]
}
```

The request was made via the fastlane [`register_devices`](https://docs.fastlane.tools/actions/register_devices/) lane. 

I'm trying to understand if this is a limitation by Apple, or if there is any known workaround.

**Comment by s6eskand** [comment]
The issue persists even with a random serial set

**Comment by s6eskand** [comment]
I was using the tart sequoia 15.1 images. I have a suspicion this is because Apple does not recognize the random UDID that tart provides in the VM, whereas the UDID of an actual mac would be verifiable, especially since you can change the UDID of the vm using `tart set <vm_name> --random-serial` so it's possible this isn't possible at the moment.

My workaround for right now has been to build apps with the `developer_id` profile since that doesn't require a registered device. More info on that [here](https://developer.apple.com/support/developer-id/) if you're curious

**Comment by CodeStage** [comment]
For signing macOS apps this is really annoying, because we have to register a new device ID for every new VM image we create. And there is a limit of 100.

I want to make a suggestion. I saw [this](https://github.com/cirruslabs/tart/pull/809/files) MR which added the --random-serial set command. 

Why not allow advanced users to pass the device ID they want to use? The [VZMacMachineIdentifier](https://developer.apple.com/documentation/virtualization/vzmacmachineidentifier) type has an initializer with Data so it should be possible in exactly the same way as with the random identifier.

Has this been discussed before? I don't want to maintain a Tart fork, but I would be willing to create a MR for it if there is a chance for it to get accepted.

**Comment by CodeStage** [comment]
No actually it's more about the device ID not changing when deploying new versions of a image. As far as I understand you can't just chose arbitrary device IDs because we don't know how VZMacMachineIdentifier works exactly. 

The documentation says "When serializing the VM to disk, you can preserve the identifier in a binary representation by serializing the data in the VZMacMachineIdentifier.[dataRepresentation](https://developer.apple.com/documentation/virtualization/vzmachardwaremodel/3816056-datarepresentation) property. Conversely, you can recreate the identifier with [init(dataRepresentation:)](https://developer.apple.com/documentation/virtualization/vzmacmachineidentifier/3816062-init) from the binary representation."

And I thought if we can just take a valid binary representation, for example when the image is first created, persist it, use it the next time we generate a new version of the image. That would ensure the device ID never changes.

**Comment by CodeStage** [comment]
Yes it changes, which confirms assumption about how code signing works. I don't know what exactly is going on, but after running `tart set --random-serial` xcodebuild will fail with the error message that the provisioning profile does not include the device and it shows the new machine identifier. Adding this new machine identifier to the developer portal fixes the issue for this image.

From playing around I can tell that you are already persisting this serial / machine identifier in some way, because merely creating a new image on the same machine will not create a new serial. In other words, every image created on the same machine always has the same serial, unless `tart set --random-serial` is run.

**Therefore my conclusion is that our issue stems from the fact that we are using a pool of several machines to create these images, and if a new machine is added to this pool, the images created on it will have a new serial.**

Which is not as bad as I thought initially, where I said the serial changes for every new image. This is not true, the serial only changes if the new image is build on a new machine.

**Comment by CodeStage** [comment]
To the original poster: I was not using fastlane to add the machine identifiers, but I am using match to generate provisioning profiles. So generally speaking there is no issue with Tart VMs and code signing. Maybe you just sent the wrong machine identifier? Where did you get your argument tart_vm_udid from? I'm taking it from the xcodebuild output...

**Comment by s6eskand** [comment]
> To the original poster: I was not using fastlane to add the machine identifiers, but I am using match to generate provisioning profiles. So generally speaking there is no issue with Tart VMs and code signing. Maybe you just sent the wrong machine identifier? Where did you get your argument tart_vm_udid from? I'm taking it from the xcodebuild output...

The issue here isn't code signing itself but code signing an app using a specific provisioning profile. The app I'm trying to build uses an iOS App Development profile, in order to build on any device using a Development profile you must register the device to the profile which returns the error I originally sent.

The `tart_vm_udid` is just the mac UDID (from the VM), grabbed by running:
```
system_profiler SPHardwareDataType | awk '/Provisioning UDID/ { print $3 }'
```
or by just checking System Information

**Comment by s6eskand** [comment]
> Are you using a macOS Sequoia VM that was built on your machine, or somewhere else?

I'm using a VM, not built on my machine, from the latest IPSW at the time (15.1). How would I build on my machine? I can try that out and see if it fixes the error

> Have you tried using the latest IPSW? For example, Sequoia 15.2 beta 3 or newer.

I haven't tried anything newer than 15.1

**Comment by s6eskand** [comment]
tried building locally and it did not resolve the issue, I believe because the hardware identifier is originally set from the restore generated and distributed by apple. If there is a way to generate the ipsw from my own mac then I think the issue may be resolved, but I can't find any information about this online.

Looking through the [virtualization docs](https://developer.apple.com/documentation/virtualization/vzmachardwaremodel) also confirms this as the hardware model settings come from the restore itself.

**Comment by protonjohn** [comment]
I am experiencing a similar issue on macOS Sequoia 15.2, where I am unable to run anything that is developer-signed, _even if_ I add the device UUID to the devices list for my organization.

Strangely, when I go to inspect the `embedded.provisionprofile` in the contents of the application I'm trying to run, the UUID is present in the list, but when I try to run the app or install the given profile, I'm given the same error that @CodeStage experienced, where the OS says that the `provisioning profile does not include the device`.

The same app will run just fine on a bare-metal machine that has been registered in the devices list. It seems that some part of Sequoia is refusing to accept any provisioning profile for running developer-signed apps, even if that profile contains the device's UUID. I'm moving forward by signing everything with Developer ID as a workaround, but obviously that isn't a very attractive solution from a security standpoint. If anybody has more details I'd love to hear your experience, so far the Internet hasn't turned up any useful information as it seems like it's quite a niche use case.

**Comment by protonjohn** [comment]
It turns out that Sequoia VMs have an incorrect format for their provisioning UDIDs. This is why you're getting that error when trying to register the VM as a device - it's using the iOS-style UDID that looks like `b043d27202c7ac37ca3c6b82673302225485cae9` when it should look more like `00001312-1234ACABDEADBEEF`. This is a known issue, but I'm not sure what their timeline is for the fix. https://developer.apple.com/forums/thread/773051

**Comment by CodeStage** [comment]
I believe Apple has changed their service to allow the new ID format
